### PR TITLE
cc,wasi: support WASI reactors via -mexec-model flag

### PR DESCRIPTION
**Pull request by mathetake**
Based on this patch to LLVM (https://reviews.llvm.org/D62922). This partially resolves (only for zig cc) #6757

### CODE REVIEW

File: src/wasi_libc.zig
Comment (kubkon) [inline src/wasi_libc.zig:66]: These changes are no longer necessary and are fixed in https://github.com/ziglang/zig/pull/8992

**Review by kubkon — CHANGES_REQUESTED** [review]
Thanks for the PR! I've got a couple of suggestions that when we address, I'm sure we'll be able to merge this change no problem.

File: src/Compilation.zig
Comment (mathetake) [reply src/Compilation.zig]: Thanks, that would be much better! Done in https://github.com/ziglang/zig/pull/8978/commits/6614a204d060411b4bbbdbcd035ccee919a5c006

File: src/wasi_libc.zig
Comment (mathetake) [reply src/wasi_libc.zig:66]: sounds good! so should I wait until that PR is merged?

**Comment by mathetake** [comment]
OK so I verified this works as expected locally.

File: src/wasi_libc.zig
Comment (kubkon) [reply src/wasi_libc.zig:66]: Actually, feel free to rebase onto my branch, and we can carry on from there. Feel free to ping me on Discord if you have any questions about the changes in that PR or whatnot!

**Comment by mathetake** [comment]
@kubkon ok now rebased and added a necessary commit for checking in command/reactor crt (https://github.com/ziglang/zig/pull/8978/commits/a2b20449efdb07ba28b739fa544d6364d6a5851c). PTAL!

File: src/Compilation.zig
Comment (kubkon) [inline src/Compilation.zig:1455]: OK, so my idea for this bit here is similar to the logic just a few lines above we use for working out which emulated WASI libc subcomponents we should build in this run (unless they are already in global cache). So, instead of building _all_ CRTs, how about we now use the new option you just added to work out precisely which CRT we should include in our build, something like:

```zig
const crt_file: wasi_libc.CRTFile = if (comp.bin_file.options.wasi_exec_model) |exec_model| crt_file: {
    switch (exec_model) {
        .command => break :crt_file .crt1_command_o,
        .reactor => break :crt_file .crt1_reactor_o,
    }
} else .crt1_o;
comp.work_queue.writeAssumeCapacity(&[_]Job{
    .{ .wasi_libc_crt_file = crt_file },
    .{ .wasi_libc_crt_file = .libc_a },
});
```

Comment (kubkon) [reply src/Compilation.zig:1455]: Also, while we're here, could you remove the TODO as it is no longer needed.

File: src/link.zig
Comment (kubkon) [inline src/link.zig:120]: Could you add a sentence or two explaining what this option holds?

File: src/Compilation.zig
Comment (kubkon) [inline src/Compilation.zig:733]: Could you add a sentence or two explaining what this option holds?

File: src/main.zig
Comment (kubkon) [inline src/main.zig:1264]: Minor nit: you actually don't need to initialize to `null` here since it is already initialized to this value by default.

File: src/link/Wasm.zig
Comment (kubkon) [inline src/link/Wasm.zig:686]: I think it is more readable to unpack the optional using the `if (opt) |value|` syntax which is more readable, and use labeled block to leave early if the inner value is not a `.reactor`:

```zig
if (self.base.options.wasi_exec_model) |exec_model| blk: {
    if (exec_model != .reactor) break :blk;
    try argv.append("--no-entry");
}
```

See [Optionals](https://ziglang.org/documentation/0.8.0/#toc-Optionals) for more details.

**Review by kubkon — CHANGES_REQUESTED** [review]

File: src/Compilation.zig
Comment (mathetake) [reply src/Compilation.zig:734]: https://github.com/ziglang/zig/pull/8978/commits/cd610975f2c401ec413787c04c2c1c8a34e54804 fixed!

File: src/link.zig
Comment (mathetake) [reply src/link.zig:121]: https://github.com/ziglang/zig/pull/8978/commits/cd610975f2c401ec413787c04c2c1c8a34e54804 fixed!

File: src/Compilation.zig
Comment (mathetake) [reply src/Compilation.zig]: fixed in https://github.com/ziglang/zig/pull/8978/commits/cd610975f2c401ec413787c04c2c1c8a34e54804 with a slight modification. I couldn't compile without explicitly giving the enum value inside of these switch branches (maybe due to type inference limitation)?

Comment (mathetake) [reply src/Compilation.zig]: Anyway thank you so much for suggestions!

File: src/main.zig
Comment (mathetake) [reply src/main.zig]: fixed in  https://github.com/ziglang/zig/pull/8978/commits/230878352006e466f66600a979bdba8bddc98acd

**Review by kubkon — APPROVED** [review]
Looks great! Thanks for working on this @mathetake!

**Comment by kubkon** [comment]
BTW once #8992 lands, I'll merge this one next.

**Comment by mathetake** [comment]
@kubkon Thanks for the review! Much appreciated, and I learned a lot from your suggestions:)

**Comment by kubkon** [comment]
OK Github, why the heck did you close this PR?!

**Comment by kubkon** [comment]
@mathetake would you mind reopening the PR?

**Comment by kubkon** [comment]
Or I can try merging your branch manually on the command line, let's see...

**Comment by mathetake** [comment]
OK, let me reopen!

**Comment by kubkon** [comment]
> OK, let me reopen!

Don't worry about conflicts, I'll take care of them while merging.

**Comment by mathetake** [comment]
https://github.com/ziglang/zig/pull/9052 here you go!

**Pull request closed without merging**
